src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g

```diff
@@ -405,18 +405,26 @@ typeArgumentBounds[boolean addImagNode]
 // A builtin type specification is a builtin type with possible brackets
 // afterwards (which would make it an array type).
 builtInTypeSpec[boolean addImagNode]
-     :    builtInType
-         (options{greedy=true; }:
-            ({LA(1) == AT}? annotations
-            | )
-                lb:LBRACK^ {#lb.setType(ARRAY_DECLARATOR);} RBRACK)*
```

Comment:
This production was creating the `TYPE` ast with the actual type of variable/ method nested within the `[]`:

code:
```java
static int[][][] method1(Object args) {}
```
resulted in:

```bash
    |   |--TYPE -> TYPE [2:18]
    |   |   `--ARRAY_DECLARATOR -> [ [2:18]
    |   |       |--ARRAY_DECLARATOR -> [ [2:16]
    |   |       |   |--ARRAY_DECLARATOR -> [ [2:14]
    |   |       |   |   |--LITERAL_INT -> int [2:11]
    |   |       |   |   `--RBRACK -> ] [2:15]
    |   |       |   `--RBRACK -> ] [2:17]
    |   |       `--RBRACK -> ] [2:19]

```
Now, we have at least extracted the type to be the first child of the  `TYPE` AST:

```bash

    |   |--TYPE -> TYPE [2:11]
    |   |   |--LITERAL_INT -> int [2:11]
    |   |   `--ARRAY_DECLARATOR -> [ [2:18]
    |   |       |--ARRAY_DECLARATOR -> [ [2:16]
    |   |       |   |--ARRAY_DECLARATOR -> [ [2:14]
    |   |       |   |   `--RBRACK -> ] [2:15]
    |   |       |   `--RBRACK -> ] [2:17]
    |   |       `--RBRACK -> ] [2:19]

```
I will open an issue to "unroll" the nested `[]` next, but this was a good first step, and fixes the issue at hand.

Reply:
Issue opened at https://github.com/checkstyle/checkstyle/issues/9928